zerver/migrations/0480_realm_backfill_uuid_and_secret.py

```diff
@@ -0,0 +1,69 @@
+import secrets
+import uuid
+
+from django.db import migrations, models
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+
+def generate_api_key() -> str:
+    """
+    This is a copy of zerver.lib.utils.generate_api_key. Importing code that's prone
+    to change in a migration is something we generally avoid, to ensure predictable,
+    consistent behavior of the migration across time.
+    """
+
+    api_key = ""
+    while len(api_key) < 32:
+        api_key += secrets.token_urlsafe(3 * 9).replace("_", "").replace("-", "")
+    return api_key[:32]
+
+
+def generate_realm_uuid_owner_secret() -> str:
+    token = generate_api_key()
+
+    return f"zuliprealm_{token}"
+
+
+def backfill_realm_uuid_and_secret(
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
+    Realm = apps.get_model("zerver", "Realm")
+
+    max_id = Realm.objects.aggregate(models.Max("id"))["id__max"]
```

Comment:
So because this migration doesn't run in a transaction, it's possible for new realms to be created while the migration is running, resulting in the last one being missed. So this isn't the ideal construction.

One way to handle things is to just go until `max_id *1.1 + 10` -- which is hacky in that there's no guarantee there aren't more realms than that, but very likely to work with minimal extra cost.

Another option, which depends on our having an index on `uuid` such that one can efficiently query for items that don't have one, is this:
```
while True:
    realms = Realm.objects.filter(uuid=None)
    if not realms.exists():
        break
    realms_to_update = realms[0:BATCH_SIZE]
    ...
```

In practice, we can manage this in Zulip Cloud because (1) the migration to remove `null=True` will fail and (2) we can run the migration again if that happens, but I figured I'd mention the problem statement since it's better to just write the migration to Just Work in the first place.

Reply:
@timabbott But when this backfill migration runs, the first commit should have already been deployed, meaning any new realms (so above the `max_id` grabbed here) will be created with the `default` value from `models.py` applied. So this should be fine? This is consistent with our strategy for other similar migration like 3e2ad84bbe7cbc8099a42b6860070cad55cb9b52. Unless I'm missing some detail here? :thinking:

Reply:
You're right that it will work in most installations, but in Zulip Cloud, we will sometimes start running a migration like this before a deployment has finished, rather than doing the different stages in separate deployments. Though I suppose in this case that's unlikely to be a thing that's useful to do, because the field won't be read, so I guess it's fine for this specific migration.

OK I'll merge this one. But for future migrations, I do very much find it easier to just try to write migrations to be robust just to have fewer possibly failure modes to think about in production deployments.